V3 fails to compile with `#[cfg(not(target_has_atomic = "ptr"))]`

When attempting to compile a crate that depends on rapidhash 3.0.0, targeting a platform with `#[cfg(not(target_has_atomic = "ptr"))]` (specifically `atmega328p`), I get the following error:

```
error[E0432]: unresolved import `crate::inner::seeding::secrets::GlobalSecrets`                                                  
 --> C:\Users\object\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\rapidhash-3.0.0\src\inner\state.rs:3:5
  |
3 | use crate::inner::seeding::secrets::GlobalSecrets;
  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ no `GlobalSecrets` in `inner::seeding::secrets`
```

The error is on this line:

https://github.com/hoxxep/rapidhash/blob/b480aed83d2b4246644ba2b96146c37aa1553eec/rapidhash/src/inner/state.rs#L3

It seems like the error occurs because the implementation of `crate::inner::seeding::secrets` for platforms without atomic pointers does not define `GlobalSecrets`.

https://github.com/hoxxep/rapidhash/blob/b480aed83d2b4246644ba2b96146c37aa1553eec/rapidhash/src/inner/seeding.rs#L72-L80

--- a/.github/workflows/rust.yaml
+++ b/.github/workflows/rust.yaml
@@ -106,3 +106,17 @@ jobs:
         run: rustup target add wasm32-unknown-unknown
       - name: Compile and run tests
         run: cargo build -p rapidhash --target wasm32-unknown-unknown
+
+  test-no-atomics:
+    # check that the no-atomics target compiles, but it sadly can't run tests
+    name: "Build (no atomics)"
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v4
+      - uses: dtolnay/rust-toolchain@stable
+      - name: Rust dependency cache
+        uses: Swatinem/rust-cache@v2
+      - name: Install no-atomics target
+        run: rustup target add thumbv6m-none-eabi
+      - name: Compile and run tests
+        run: cargo build -p rapidhash --no-default-features --lib --target thumbv6m-none-eabi

--- a/rapidhash/src/inner/seeding.rs
+++ b/rapidhash/src/inner/seeding.rs
@@ -44,6 +44,9 @@ pub(crate) mod seed {
 
         // without std we fall back to a global atomic and accept the chance of
         // race conditions, but don't consider this an issue
+        //
+        // Most targets without atomics can still do atomic load/store, but just can't
+        // do atomic compare-and-swap instructions. So this should still compile/work...
         #[cfg(not(feature = "std"))] {
             use core::sync::atomic::{AtomicUsize, Ordering};
             static RANDOM_SEED: AtomicUsize = AtomicUsize::new(0);
@@ -76,6 +79,28 @@ pub(crate) mod secrets {
         // This is a no-op for platforms that do not support atomic pointers.
         // The secrets are not used, so we return an empty slice.
         &crate::inner::seed::DEFAULT_RAPID_SECRETS.secrets
+    }
+
+    #[derive(Copy, Clone, Debug, Eq, PartialEq)]
+    pub struct GlobalSecrets {
+        _only_uses_default_secrets: (),
+    }
+
+    impl GlobalSecrets {
+        /// Set up the global secrets if they are not already initialized.
+        #[inline(always)]
+        pub fn new() -> Self {
+            Self {
+                _only_uses_default_secrets: (),
+            }
+        }
+
+        /// Get the global secrets, which are guaranteed to be initialized, but these will
+        /// be the default rapidhash secrets as this target does not support atomic pointers.
+        #[inline(always)]
+        pub fn get(self) -> &'static [u64; 7] {
+            get_secrets()
+        }
     }
 }
 

--- a/rapidhash/src/inner/state.rs
+++ b/rapidhash/src/inner/state.rs
@@ -8,6 +8,17 @@ use crate::inner::seeding::secrets::GlobalSecrets;
 /// This is designed to provide some HashDoS resistance by using a random seed per hashmap, and
 /// a global random set of secrets.
 ///
+/// # Portability
+///
+/// On most target platforms, the secrets are randomly initialized once and cached globally for the
+/// lifetime of the program using a mix of ASLR and other entropy sources. The seed is randomly
+/// initialized for each new instance of `RandomState` using only ASLR and a mixing step.
+///
+/// On targets without atomic pointer support, the global secrets will not be randomised, and
+/// instead will fall back to the default secrets. This means these platforms will not have minimal
+/// HashDoS resistance guarantees. If this is important for your application, please raise a GitHub
+/// issue to improve support for these platforms.
+///
 /// # Example
 /// ```rust
 /// use std::collections::HashMap;

--- a/rapidhash/src/inner/state.rs
+++ b/rapidhash/src/inner/state.rs
@@ -29,12 +29,17 @@ pub struct RandomState<const AVALANCHE: bool, const SPONGE: bool, const COMPACT:
 impl<const AVALANCHE: bool, const SPONGE: bool, const COMPACT: bool, const PROTECTED: bool> RandomState<AVALANCHE, SPONGE, COMPACT, PROTECTED> {
     /// Create a new random state with a random seed.
     ///
-    /// With the `rand` feature enabled, this will use [rand::random] to initialise the seed.
+    /// The seed is always randomised by using ASLR on every new instance of RandomState.
     ///
-    /// Without `rand` but with the `std` feature enabled, this will use [crate::rapidrng_time] to
-    /// initialise the seed.
+    /// With the `rand` feature enabled, the secrets will be randomised using [rand::random].
+    /// Otherwise, a mix of ASLR and some other poorer sources of entropy will be mixed together to
+    /// generate the secrets. The secrets are statically cached for the lifetime of the program
+    /// after their initial generation.
+    ///
+    /// On platforms that do not support atomic pointers, the secrets will be the default rapidhash
+    /// secrets, which are not randomised. Therefore, **targets without atomic pointer support will
+    /// not have minimal HashDoS resistance guarantees**.
     #[inline]
-    #[cfg(target_has_atomic = "ptr")]
     pub fn new() -> Self {
         Self {
             seed: super::seeding::seed::get_seed(),

